zug/util.hpp

```diff
@@ -52,48 +57,63 @@ ZUG_INLINE_CONSTEXPR struct identity__t
 
 namespace detail {
 
-template <class F, class G>
+template <typename Args, typename Fn1, typename Fn2>
+constexpr decltype(auto) apply_composition(Args&& args, Fn1&& fn1, Fn2&& fn2)
+{
+    return compat::invoke(
+        std::forward<Fn1>(fn1),
+        compat::apply(std::forward<Fn2>(fn2), std::forward<Args>(args)));
+}
+
+template <typename Args, typename Fn, typename... Fns, typename... T>
+constexpr decltype(auto) apply_composition(Args&& args, Fn&& fn, Fns&&... fns)
+{
+    return compat::invoke(
+        std::forward<Fn>(fn),
+        apply_composition(std::forward<Args>(args), std::forward<Fns>(fns)...));
+}
+
+template <typename Fns, typename Args, size_t... Is>
+constexpr decltype(auto)
+apply_composition(Fns&& fns, Args&& args, std::index_sequence<Is...>)
+{
+    return apply_composition(std::forward<Args>(args),
+                             std::get<Is>(std::forward<Fns>(fns))...);
+}
+
+template <typename Fns, typename Args>
+constexpr decltype(auto) apply_composition(Fns&& fns, Args&& args)
+{
+    return apply_composition(
+        std::forward<Fns>(fns),
+        std::forward<Args>(args),
+        std::make_index_sequence<std::tuple_size<std::decay_t<Fns>>::value>{});
+}
+
+template <typename Fn, typename... Fns>
 struct composed
 {
-    F f;
-    G g;
+    std::tuple<Fn, Fns...> fns;
 
-    template <class... T>
-    decltype(auto) operator()(T&&... xs) &
+    template <typename... T>
+    constexpr decltype(auto) operator()(T&&... xs) &
     {
-        return compat::invoke(f, compat::invoke(g, std::forward<T>(xs)...));
+        return apply_composition(fns, std::forward_as_tuple(xs...));
```

Comment:
I'm not fan of using `forward_as_tuple` for the args here. I understand you did this to be able to expand all the functions in the argument list and distinguish them from the args. Instead of expanding all the functions in the argument list and then re-expanding the arguments at the end, I would just pick the functions one by one, and keep the final arguments expanded/forwarded at all times. You do not need in that case an index sequence, just need to carry a template argument with the current index in the functions tuple, and compare in each step with the size of the tuple. Since we are on C++17 you can use `if constexpr` to avoid SFINAE for that.

Reply:
We are still on C++14! But point taken...